lua/ttt2/libraries/migrations.lua

```diff
@@ -0,0 +1,184 @@
+migrations.isLoaded = false
+
+---
+-- Get ORM table and create it if not available yet
+-- @return @{ORMOBJECT} The orm-table to address the database
+-- @realm shared
+function migrations.GetORM()
+	if istable(migrations.orm) then
+		return migrations.orm
+	end
+
+	-- Create Sql and orm table if not already done or savingKeys were changed
+	sql.CreateSqlTable(migrations.databaseName, migrations.savingKeys)
+	migrations.orm = orm.Make(migrations.databaseName)
+
+	return migrations.orm
+end
+
+---
+-- Creates a command with state information, an up- and downgrade function
+-- @param table states The states that are needed to fulfill the upgrade
+-- @param function upgrade The function that is used for forward operations
+-- @param function downgrade The function that is used for reverting operations of this command
+-- @return @{command} A command containing necessary functionality for forward and backward operations
+-- @realm shared
+function migrations.CreateCommand(states, upgrade, downgrade)
+	if not istable(states) or not isfunction(upgrade) or not isfunction(downgrade) then
+		ErrorNoHalt("[TTT2] Couldn't create migrations command. Missing states-table, upgrade- or downgrade-function.\n")
+		states = {}
+		upgrade = function(cmd) return end
+		downgrade = function(cmd) return end
+	end
+
+	local command = {Upgrade = upgrade, Downgrade = downgrade, isCommand = true}
+	table.Merge(command, states)
+
+	return command
+end
+
+---
+-- Adds a migration command, that is executed on migration
+-- @Note Only used in `terrortown/migrations` and corresponding `client`/`server`/`shared` folders
+-- @param @{command} The command to execute on migration
+-- @return bool True, if adding was successful
+-- @realm shared
+function migrations.Add(command)
+	if not istable(command) or not command.isCommand then
+		ErrorNoHalt("[TTT2] Couldn't add migration command. Missing version or command.\n")
+		return false
+	end
+
+	migrations.cachedCommands[#migrations.cachedCommands + 1] = command
+
+	return true
+end
+
+---
+-- Runs all missing forward gamemode migrations.
+-- @return boolean `true` if the desired version was successfully migrated and `false` in case of an error.
+-- @realm shared
+function migrations.Apply()
+	if not migrations.isLoaded then
+		migrations.Load()
+	end
+
+	local migrationSuccess = true
+	local errorMessage = ""
+	local databaseORM = migrations.GetORM()
+
+	for fileName, commands in SortedPairs(migrations.commands) do
+		local fileInfo = databaseORM:Find(fileName)
+
+		-- As ORM currently gives out strings check it
+		if fileInfo and (fileInfo.wasSuccess == true or fileInfo.wasSuccess == "true") then continue end
+
+		MsgN("[TTT2] Migrating: ", fileName)
+
+		for j = 1, #commands do
+			migrationSuccess, errorMessage = pcall(commands[j].Upgrade, commands[j])
+
+			if not migrationSuccess then break end
+		end
+
+		local date = os.date("%Y/%m/%d - %H:%M:%S", os.time())
+
+		if fileInfo then
+			fileInfo.date = date
```

Comment:
date isn't really describing what this information is. something like `executed_at` / `executedAt` might be a better description for this (especially in the database)